### ISSUE TITLE
OutOfMemoryError on Event Replay

### ISSUE THREAD

**Issue by nattyddubbs** [RBMHTechnology/eventuate#268]
During event replay I ran into the following exception:

```
2016-06-06 08:06:28.061 [application-akka.actor.default-dispatcher-2] ERROR akka.actor.ActorSystemImpl - Uncaught fatal error from thread [application-akka.actor.default-dispatcher-2] shutting down ActorSystem [application]
java.lang.OutOfMemoryError: Java heap space
        at com.google.protobuf.AbstractMessageLite.toByteArray(AbstractMessageLite.java:64) ~[protobuf-java-3.0.0-beta-2.jar:na]
        at com.rbmhtechnology.eventuate.serializer.DurableEventSerializer.toBinary(DurableEventSerializer.scala:43) ~[eventuate-core_2.11-0.7.1.jar:0.7.1]
        at akka.serialization.Serialization$$anonfun$serialize$1.apply(Serialization.scala:91) ~[akka-actor_2.11-2.4-bin-rp-16s01p03.jar:na]
        at akka.serialization.Serialization$$anonfun$serialize$1.apply(Serialization.scala:91) ~[akka-actor_2.11-2.4-bin-rp-16s01p03.jar:na]
        at scala.util.Try$.apply(Try.scala:192) ~[scala-library-2.11.7.jar:na]
        at akka.serialization.Serialization.serialize(Serialization.scala:91) ~[akka-actor_2.11-2.4-bin-rp-16s01p03.jar:na]
        at com.rbmhtechnology.eventuate.log.cassandra.Cassandra.eventToByteBuffer(Cassandra.scala:217) ~[eventuate-log-cassandra_2.11-0.7.1.jar:0.7.1]
        at com.rbmhtechnology.eventuate.log.cassandra.CassandraIndexStore$$anonfun$writeAggregateEventsAsync$3$$anonfun$apply$2.apply(CassandraIndexStore.scala:57) ~[eventuate-log-cassandra_2.11-0.7.1.jar:0.7.1]
        at com.rbmhtechnology.eventuate.log.cassandra.CassandraIndexStore$$anonfun$writeAggregateEventsAsync$3$$anonfun$apply$2.apply(CassandraIndexStore.scala:57) ~[eventuate-log-cassandra_2.11-0.7.1.jar:0.7.1]
        at scala.collection.Iterator$class.foreach(Iterator.scala:742) ~[scala-library-2.11.7.jar:na]
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1194) ~[scala-library-2.11.7.jar:na]
        at scala.collection.IterableLike$class.foreach(IterableLike.scala:72) ~[scala-library-2.11.7.jar:na]
        at scala.collection.AbstractIterable.foreach(Iterable.scala:54) ~[scala-library-2.11.7.jar:na]
        at com.rbmhtechnology.eventuate.log.cassandra.CassandraIndexStore$$anonfun$writeAggregateEventsAsync$3.apply(CassandraIndexStore.scala:57) ~[eventuate-log-cassandra_2.11-0.7.1.jar:0.7.1]
        at com.rbmhtechnology.eventuate.log.cassandra.CassandraIndexStore$$anonfun$writeAggregateEventsAsync$3.apply(CassandraIndexStore.scala:56) ~[eventuate-log-cassandra_2.11-0.7.1.jar:0.7.1]
        at com.rbmhtechnology.eventuate.log.cassandra.Cassandra.withBatch(Cassandra.scala:239) ~[eventuate-log-cassandra_2.11-0.7.1.jar:0.7.1]
        at com.rbmhtechnology.eventuate.log.cassandra.Cassandra.executeBatchAsync(Cassandra.scala:235) ~[eventuate-log-cassandra_2.11-0.7.1.jar:0.7.1]
        at com.rbmhtechnology.eventuate.log.cassandra.CassandraIndexStore.writeAggregateEventsAsync(CassandraIndexStore.scala:56) ~[eventuate-log-cassandra_2.11-0.7.1.jar:0.7.1]
        at com.rbmhtechnology.eventuate.log.cassandra.CassandraIndexStore$$anonfun$writeAggregateEventsAsync$1.apply(CassandraIndexStore.scala:53) ~[eventuate-log-cassandra_2.11-0.7.1.jar:0.7.1]
        at com.rbmhtechnology.eventuate.log.cassandra.CassandraIndexStore$$anonfun$writeAggregateEventsAsync$1.apply(CassandraIndexStore.scala:52) ~[eventuate-log-cassandra_2.11-0.7.1.jar:0.7.1]
        at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:245) ~[scala-library-2.11.7.jar:na]
        at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:245) ~[scala-library-2.11.7.jar:na]
        at scala.collection.immutable.HashMap$HashMap1.foreach(HashMap.scala:221) ~[scala-library-2.11.7.jar:na]
        at scala.collection.immutable.HashMap$HashTrieMap.foreach(HashMap.scala:428) ~[scala-library-2.11.7.jar:na]
        at scala.collection.immutable.HashMap$HashTrieMap.foreach(HashMap.scala:428) ~[scala-library-2.11.7.jar:na]
        at scala.collection.immutable.HashMap$HashTrieMap.foreach(HashMap.scala:428) ~[scala-library-2.11.7.jar:na]
        at scala.collection.TraversableLike$class.map(TraversableLike.scala:245) ~[scala-library-2.11.7.jar:na]
        at scala.collection.AbstractTraversable.map(Traversable.scala:104) ~[scala-library-2.11.7.jar:na]
        at com.rbmhtechnology.eventuate.log.cassandra.CassandraIndexStore.writeAggregateEventsAsync(CassandraIndexStore.scala:52) ~[eventuate-log-cassandra_2.11-0.7.1.jar:0.7.1]
        at com.rbmhtechnology.eventuate.log.cassandra.CassandraIndexStore.writeAsync(CassandraIndexStore.scala:46) ~[eventuate-log-cassandra_2.11-0.7.1.jar:0.7.1]
        at com.rbmhtechnology.eventuate.log.cassandra.CassandraIndexUpdater.writeAsync(CassandraIndex.scala:126) ~[eventuate-log-cassandra_2.11-0.7.1.jar:0.7.1]
        at com.rbmhtechnology.eventuate.log.cassandra.CassandraIndexUpdater$$anonfun$updateAsync$1.apply(CassandraIndex.scala:122) ~[eventuate-log-cassandra_2.11-0.7.1.jar:0.7.1]
```

As you can see we are using the latest version of Eventuate. If I move back to version 6.0 I also run into this problem.

This error did not occur while persisting the events to the event log however it seems to occur when replaying. I'm wondering what could be causing this.

Nathaniel

**Comment by nattyddubbs** [comment]
More information:

- This is a one node cluster
- Replay is reprocessing 300k + events.

**Comment by krasserm** [comment]
- What are the JVM settings (heap size, ...)?
- What are the custom `eventuate.*` configuration settings?
- How many `Eventsourced*` actors do you have?
- How many of them are updated simultaneously?
- How many different `aggregateId`s are used?
- What's the usage (read/write load) profile of these actors?
- What's the size of your application-level event objects (i.e. `DurableEvent` payloads)?
- How many event log actors are instantiated?
- Have you read the [failure handling](http://rbmhtechnology.github.io/eventuate/reference/event-sourcing.html#failure-handling) section regarding write batch sizes
- ...
- Can you provide a failing test?

**Comment by nattyddubbs** [comment]
Just to reiterate this error occurs when the service is started. It is not actively servicing any requests or generating any events while this error occurs. 

**What are the JVM settings (heap size, ...)?**: Xms: 512, Xmx:4096

**What are the custom `eventuate`.* configuration settings**: 
eventuate {
  log {
    # Timeout for read operations from a local event log. These are batch
    # event reads during event replay made by event-sourced views, actors,
    # writers and processors and target log reads made by processors.
    read-timeout = 10s

    # Maximum number of events to be replayed to an event-sourced view, actor,
    # writer or processor before replay is suspended. A suspended replay is
    # resumed automatically after the replayed event batch has been handled
    # (= replay backpressure).
    replay-batch-size = 4000

    cassandra {
      # Comma-separated list of contact points in the cluster (comma-separated
      # hostname or hostname:port list).
      contact-points = ["cassandra1", "cassandra2"]

      # Default port of contact points in the cluster. Ports defined in
      # contact-points override this setting.
      default-port = ${?cassandra.port}

      # Name of the keyspace created/used by Eventuate.
      keyspace = "eventuate_talentreef"
    }
  }
  snapshot.filesystem {
    # Root directory for storing the snapshot files of individual emitters.
    dir = "<our own dir>"

    # Maximum number of stored snapshots per emitter. If this number is
    # exceeded during a snapshot write, older snapshots are automatically
    # deleted.
    snapshots-per-emitter-max = 3
  }
}

**How many Eventsourced* actors do you have?**: There are 16 that we start when the application starts. During normal service execution there can be upwards of 100k actors running.

 **How many of them are updated simultaneously?**: During event replay none of them are updated simultaneously. We start the actors individually and wait for them to complete replay prior to starting the next.

**How many different aggregateIds are used?**: 340k

**What's the usage (read/write load) profile of these actors?**: I'm not sure what you mean. Messages per second?

**What's the size of your application-level event objects (i.e. DurableEvent payloads)?**: At most 5k

**How many event log actors are instantiated?**: 1

**Have you read the failure handling section regarding write batch sizes?**: Yes. Our problem pertains specifically to the replay of events without writes occurring.

**Can you provide a failing test?**: No. I can look through the test cases provided in the framework and see if any can be augmented to replicate by situation.

**Comment by krasserm** [comment]
Thanks for the detailed answer. Does it help when you set `eventuate.log.cassandra.index-update-limit` to a smaller value e.g. `32` or `16`?

**Comment by nattyddubbs** [comment]
It does not. I attempted to use 16 but that resulted in the same error.

**Comment by krasserm** [comment]
@nattyddubbs unfortunately, I'm unable to reproduce. Please provide a failing test so that I can further investigate. Thanks!

**Comment by nattyddubbs** [comment]
@krasserm I tracked down the issue. It had to do with how we were using custom aggregate ids during event persistence. I appreciate you taking the time to respond to my questions.

**Comment by magro** [comment]
@nattyddubbs Is there anything you learned one should be aware of when using aggregateIds that's not covered by the docs?

**Comment by nattyddubbs** [comment]
@magro I'll write up some information regarding our use case and lessons learned and leave it here if that's acceptable.

**Comment by magro** [comment]
Sounds good!

**Comment by nattyddubbs** [comment]
We have three domain entities call them A, B and C. There is a one to many composition relationship between A and B with C attached to this relationship (we'll call this D). We have 100s of As, 10,000s of Bs, 10s of Cs and therefore 1M+ D. 

We implemented A, B, C as well as the relationship, D, as `EventsourcedActors`. A, B, C and D each have unique aggregate ids on a per actor instance basis. However, logic within D requires that it be notified whenever an event occurs on C. The way we accomplished this is by using the custom aggregate id feature in the `onEvent` call.

During application initialization, we run several scripts that ensure all of C exist and are up to date. These values change infrequently and therefore should not cause a lot of events to be generated. HOWEVER our logic was bad and events were being generated every time the application started. The number of events was on the magnitude of 10 but for each event there were 20k+ custom aggregate ids on each onEvent call.

This 100k+ events to be serialized and written to the aggregate event log. Serialization was eating up all of our physical memory as Cassandra could not 'keep up' because of the volume of writes so the batches were sitting in memory in our service.

Our fixes were
1. Fix the diff logic to decide events for C
2. Remove the custom aggregate id persistence in onEvent
3. Do a client side join of the data that composes D instead of having D be updated whenever C has events.

I hope that this makes sense and is helpful. If not I can elaborate and clarify as appropriate.
